logger_test.go

```diff
@@ -42,17 +46,65 @@ func (tw *testOutput) Write(p []byte) (int, error) {
 
 func (tw *testOutput) Sync() error { return nil }
 
-func (tw *testOutput) requireContains(t *testing.T, keys []string) {
+func (tw *testOutput) validate(t *testing.T, keys []string) {
 	for _, s := range keys {
 		require.Contains(t, tw.m, s)
 	}
+
+	// Fields `log.origin.file.line` and `log.origin.file.name` are logged as
+	// a map under key log.origin. By using the zap jsonEncoder this cannot
+	// be changed. Remove the nested and add a dotted version of the fields
+	if caller, ok := tw.m[callerKey].(map[string]interface{}); ok {
+		for name, val := range caller {
+			tw.m[fmt.Sprintf("%s.%s", callerKey, name)] = val
+			delete(tw.m, name)
+		}
+	}
+	// skip index checks as we do not have control over that with the formatter
+	// skip Default value checks as they are not yet implemented
+	// skip TopLevelField check as ecszap logger logs in dot notation anyways
+	// skip Sanitization as it is not implemented yet
+	for name, val := range tw.m {
+		field, ok := spec.V1.Fields[name]
+		if !ok { // custom field not defined in spec
+			continue
+		}
+		if field.Required { // all required fields must be present in the log line
+			require.Contains(t, tw.m, name)
+			require.NotNil(t, tw.m[name])
+		}
```

Comment:
This will never fail, because we're ranging over `tw.m` rather than `spec.V1.Fields`. We should swap that around.

Reply:
good catch - updated